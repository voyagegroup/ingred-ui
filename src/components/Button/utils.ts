import { Theme } from "../../themes";
import { ButtonColor, ButtonColorStyle } from "./Button";

export const getContainerColorStyles = (
  theme: Theme,
): { [P in ButtonColor]: ButtonColorStyle } => ({
  primary: {
    normal: {
      background: theme.palette.primary.main,
      color: theme.palette.text.white,
      boxShadow: theme.shadow["3dShadowPrimary"],
      border: `1px solid ${theme.palette.primary.dark}`,
    },
    hover: {
      background: theme.palette.primary.dark,
      border: "none",
    },
    active: {
      background: theme.palette.primary.dark,
      boxShadow: theme.shadow["3dShadowActive"],
      border: "none",
    },
  },
  basicLight: {
    normal: {
      background: theme.palette.background.default,
      color: theme.palette.black,
      boxShadow: theme.shadow["3dShadowBasic"],
      border: `1px solid ${theme.palette.divider}`,
    },
    hover: {
      background: theme.palette.gray.highlight,
      border: `1px solid ${theme.palette.divider}`,
    },
    active: {
      background: theme.palette.gray.highlight,
      boxShadow: theme.shadow["3dShadowActive"],
      border: `1px solid ${theme.palette.divider}`,
    },
  },
  basicDark: {
    normal: {
      background: theme.palette.basicDark.main,
      color: theme.palette.black,
      boxShadow: theme.shadow["3dShadowBasic"],
      border: `1px solid ${theme.palette.basicDark.deepDark}`,
    },
    hover: {
      background: theme.palette.basicDark.dark,
      border: "none",
    },
    active: {
      background: theme.palette.basicDark.dark,
      boxShadow: theme.shadow["3dShadowBasicActive"],
      border: "none",
    },
  },
  danger: {
    normal: {
      background: theme.palette.danger.main,
      color: theme.palette.text.white,
      boxShadow: theme.shadow["3dShadowDanger"],
      border: `1px solid ${theme.palette.danger.dark}`,
    },
    hover: {
      background: theme.palette.danger.dark,
      border: "none",
    },
    active: {
      background: theme.palette.danger.dark,
      boxShadow: theme.shadow["3dShadowActive"],
      border: "none",
    },
  },
  clear: {
    normal: {
      background: "none",
      color: theme.palette.black,
      boxShadow: "none",
      border: "none",
    },
    hover: {
      background: theme.palette.gray.light,
      border: "none",
    },
    active: {
      background: theme.palette.gray.main,
      boxShadow: "none",
      border: "none",
    },
  },
  primaryPale: {
    normal: {
      background: theme.palette.primaryPale.main,
      color: theme.palette.primaryPale.dark,
      boxShadow: theme.shadow["3dShadowPrimaryPale"],
      border: `1px solid ${theme.palette.primaryPale.light}`,
    },
    hover: {
      background: theme.palette.primaryPale.highlight,
      border: "none",
    },
    active: {
      background: theme.palette.primaryPale.highlight,
      boxShadow: theme.shadow["3dShadowPrimaryPaleActive"],
      border: "none",
    },
  },
});
