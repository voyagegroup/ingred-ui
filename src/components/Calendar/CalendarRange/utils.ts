import dayjs, { Dayjs } from "dayjs";

const isStart = (startDate: Dayjs | null, month: Dayjs, day: number) =>
  startDate?.format("YYYY-MM-DD") ===
  dayjs(new Date(month.year(), month.month(), day)).format("YYYY-MM-DD");

const isEnd = (endDate: Dayjs | null, month: Dayjs, day: number) =>
  endDate?.format("YYYY-MM-DD") ===
  dayjs(new Date(month.year(), month.month(), day)).format("YYYY-MM-DD");

const isBetween = (
  startDate: Dayjs | null,
  endDate: Dayjs | null,
  month: Dayjs,
  day: number,
) =>
  (startDate &&
    endDate &&
    dayjs(new Date(month.year(), month.month(), day)).isAfter(
      startDate.format("YYYY-MM-DD"),
      "day",
    ) &&
    dayjs(new Date(month.year(), month.month(), day)).isBefore(
      endDate.format("YYYY-MM-DD"),
      "day",
    )) ??
  false;

export const getDayState = (
  startDate: Dayjs | null,
  endDate: Dayjs | null,
  month: Dayjs,
  day: number,
) => {
  if (isStart(startDate, month, day)) {
    return "start";
  }
  if (isEnd(endDate, month, day)) {
    return "end";
  }
  if (isBetween(startDate, endDate, month, day)) {
    return "between";
  }
  return "none";
};
