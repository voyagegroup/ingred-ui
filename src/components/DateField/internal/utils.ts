export const getInputWidth = (input: string) =>
  [...input.split("")].reduce((acc, char) => {
    const code = char.charCodeAt(0);
    return acc + (code >= 0x3000 && code <= 0xff60 ? 2 : 1);
  }, 0);
