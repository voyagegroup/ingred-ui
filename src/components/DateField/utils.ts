import { useMemo } from "react";

type Sections = {
  start: number;
  end: number;
  value: string;
  editable: boolean;
};

/**
 * 何らかのフォーマットで入ってくる日付を開始位置と終了位置と値を持つセクションに分割する
 * useDateField で format された日付操作を汎用的に行うために必要なプロパティを返す
 * 例) 2023-01-02 -> [
 *   { start: 0, end: 3, value: "2023", editable: true },
 *   { start: 4, end: 4, value: "-", editable: false },
 *   { start: 5, end: 6, value: "01", editable: true },
 *   { start: 7, end: 7, value: "-", editable: false },
 *   { start: 8, end: 9, value: "02", editable: true }
 * ]
 *
 * @param formattedDate 何らかのフォーマットで入ってくる日付
 * @returns 開始位置と終了位置と値を持つセクション
 */
export const getSections = (formattedDate?: string | null) => {
  if (!formattedDate) {
    return [];
  }

  const sections: Sections[] = [];
  let start = 0;
  let isPrevCharDigit = !isNaN(Number(formattedDate[0]));

  for (let index = 1; index <= formattedDate.length; index++) {
    const currentChar = formattedDate[index];
    const isCurrentCharDigit =
      !isNaN(Number(currentChar)) && currentChar !== " ";

    if (
      isCurrentCharDigit !== isPrevCharDigit ||
      currentChar === " " ||
      index === formattedDate.length
    ) {
      sections.push({
        start,
        end: index - 1,
        value: formattedDate.slice(start, index),
        editable: isPrevCharDigit,
      });
      start = index;
      isPrevCharDigit = isCurrentCharDigit;
    }
  }

  return sections;
};

/**
 * 開始位置と終了位置と値を持つセクションをフォーマットされた日付に変換する
 */
export const formatString = (sectionsWithCharactor: Sections[]) =>
  sectionsWithCharactor.map((section) => section.value).join("");

type ReactRef<T> =
  | React.RefCallback<T>
  | React.MutableRefObject<T>
  | React.ForwardedRef<T>
  | string
  | null
  | undefined;

// from: https://github.com/voyagegroup/ingred-ui/blob/master/src/hooks/useMergeRefs.ts
export function useMergeRefs<T>(...refs: ReactRef<T>[]): React.Ref<T> {
  return useMemo(() => {
    if (refs.every((ref) => ref === null)) {
      return null;
    }
    return (refValue: T) => {
      for (const ref of refs) {
        if (typeof ref === "function") {
          ref(refValue);
        } else if (ref && typeof ref !== "string") {
          ref.current = refValue;
        }
      }
    };
  }, [refs]);
}
