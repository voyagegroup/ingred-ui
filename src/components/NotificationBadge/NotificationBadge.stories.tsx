import * as React from "react";
import NotificationBadge, { NotificationBadgeProps } from ".";
import { Icon } from "..";
import { Story } from "@storybook/react/types-6-0";

export default {
  title: "Components/Data Display/NotificationBadge",
  component: NotificationBadge,
};

export const Number: Story<NotificationBadgeProps> = (args) => (
  <NotificationBadge {...args}>
    <Icon name="setting" type="fill" size="lg" />
  </NotificationBadge>
);

Number.args = {
  badgeContent: 100,
};

export const Text: Story<NotificationBadgeProps> = (args) => (
  <NotificationBadge {...args}>
    <Icon name="setting" type="fill" size="lg" />
  </NotificationBadge>
);

Text.args = {
  badgeContent: "New",
};

export const Dot: Story<NotificationBadgeProps> = (args) => (
  <NotificationBadge {...args}>
    <Icon name="setting" type="fill" size="lg" />
  </NotificationBadge>
);

Dot.args = {
  variant: "dot",
};
