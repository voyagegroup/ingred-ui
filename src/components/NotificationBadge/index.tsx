export { default } from "./NotificationBadge";
export type { NotificationBadgeProps } from "./NotificationBadge";
