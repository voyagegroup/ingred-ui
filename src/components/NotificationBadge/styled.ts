import styled from "styled-components";

export const Container = styled.div`
  position: relative;
  display: inline-flex;
  flex-shrink: 0;
  vertical-align: middle;
`;

type BadgeProps = {
  variant: "dot" | "normal";
  position: string;
  size: string;
  invisible: boolean;
};

const dotSizeMapping: { [size in string]: string } = {
  small: "8px",
  medium: "12px",
  large: "16px",
};

export const Badge = styled.span<BadgeProps>`
  position: absolute;
  top: ${({ position }) =>
    ["top-right", "top-left"].includes(position) ? 0 : "auto"};
  bottom: ${({ position }) =>
    ["bottom-right", "bottom-left"].includes(position) ? 0 : "auto"};
  right: ${({ position }) =>
    ["top-right", "bottom-right"].includes(position) ? 0 : "auto"};
  left: ${({ position }) =>
    ["top-left", "bottom-left"].includes(position) ? 0 : "auto"};
  transform: ${({ invisible }) => (invisible ? "scale(0)" : "scale(1)")}
    translate(
      ${({ position }) =>
        ["top-right", "bottom-right"].includes(position) ? "50%" : "-50%"},
      ${({ position }) =>
        ["bottom-right", "bottom-left"].includes(position) ? "50%" : "-50%"}
    );
  transform-origin: ${({ position }) =>
    `${["top-right", "bottom-right"].includes(position) ? "100%" : "0%"} ${
      ["bottom-right", "bottom-left"].includes(position) ? "100%" : "0%"
    }`};
  display: flex;
  align-items: center;
  align-content: center;
  justify-content: center;
  height: ${({ variant, size }) =>
    variant === "normal" ? "20px" : dotSizeMapping[size]};
  min-width: ${({ variant, size }) =>
    variant === "normal" ? "20px" : dotSizeMapping[size]};
  line-height: 1;
  padding: 0 ${({ variant }) => (variant === "normal" ? "6px" : 0)};
  border-radius: 10rem;
  color: ${({ theme }) => theme.palette.text.white};
  background-color: ${({ theme }) => theme.palette.danger.main};
  font-size: 10.5px;
  font-weight: bold;
  transition: transform 0.3s;
`;
