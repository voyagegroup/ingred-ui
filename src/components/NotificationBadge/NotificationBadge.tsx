import * as React from "react";
import * as Styled from "./styled";

export type NotificationBadgeProps = {
  variant?: "normal" | "dot";
  badgeContent?: number | string;
  position?: "top-right" | "top-left" | "bottom-right" | "bottom-left";
  max?: number;
  showZero?: boolean;
  invisible?: boolean;
  dotSize?: "small" | "medium" | "large";
  children: React.ReactNode;
};

const NotificationBadge = React.forwardRef<
  HTMLDivElement,
  NotificationBadgeProps
>(function NotificationBadge(
  {
    variant = "normal",
    badgeContent = "",
    position = "top-right",
    dotSize = "medium",
    max = 99,
    showZero = false,
    invisible: invisibleProp = false,
    children,
  },
  ref,
) {
  let invisible = invisibleProp;

  if (
    (badgeContent === 0 && !showZero) ||
    (badgeContent === "" && variant !== "dot")
  ) {
    invisible = true;
  }

  let displayValue: number | string = "";

  if (variant !== "dot") {
    if (typeof badgeContent === "number") {
      displayValue = badgeContent > max ? `${max}+` : badgeContent;
    } else {
      displayValue = badgeContent;
    }
  }

  return (
    <Styled.Container ref={ref}>
      {children}
      <Styled.Badge
        variant={variant}
        position={position}
        size={dotSize}
        invisible={invisible}
      >
        {displayValue}
      </Styled.Badge>
    </Styled.Container>
  );
});

export default NotificationBadge;
