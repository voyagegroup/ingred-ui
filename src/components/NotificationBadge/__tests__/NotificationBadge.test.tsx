import * as React from "react";
import "@testing-library/jest-dom/extend-expect";
import { cleanup } from "@testing-library/react";
import NotificationBadge from "..";
import { renderWithThemeProvider } from "../../../utils/renderWithThemeProvider";

describe("NotificationBadge component testing", () => {
  afterEach(cleanup);

  test("NotificationBadge normal", () => {
    const { asFragment } = renderWithThemeProvider(
      <NotificationBadge variant="normal" badgeContent={1}>
        text
      </NotificationBadge>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test("NotificationBadge dot", () => {
    const { asFragment } = renderWithThemeProvider(
      <NotificationBadge variant="dot">text</NotificationBadge>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
