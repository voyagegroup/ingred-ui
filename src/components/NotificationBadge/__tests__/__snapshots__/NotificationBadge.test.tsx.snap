// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotificationBadge component testing NotificationBadge dot 1`] = `
<DocumentFragment>
  <div
    class="sc-bdvvtL jQVSpr"
  >
    text
    <span
      class="sc-gsDKAQ hmTpyD"
    />
  </div>
</DocumentFragment>
`;

exports[`NotificationBadge component testing NotificationBadge normal 1`] = `
<DocumentFragment>
  <div
    class="sc-bdvvtL jQVSpr"
  >
    text
    <span
      class="sc-gsDKAQ ScVXF"
    >
      1
    </span>
  </div>
</DocumentFragment>
`;
